src/Sarif.Multitool/Rules/SARIF2004.OptimizeFileSize.cs

```diff
@@ -41,20 +41,107 @@ public class OptimizeFileSize : SarifValidationSkimmerBase
+        /// analyzer that is instructed to analyze example.c, and detects a result in the included file
+        /// example.h. In this case, 'analysisTarget' is example.c, and the result location is in example.h.
         /// </summary>
         public override MultiformatMessageString FullDescription => new MultiformatMessageString { Text = RuleResources.SARIF2004_OptimizeFileSize_FullDescription_Text };
 
         protected override IEnumerable<string> MessageResourceNames => new string[] {
+            nameof(RuleResources.SARIF2004_OptimizeFileSize_Warning_AvoidDuplicativeAnalysisTarget_Text),
+            nameof(RuleResources.SARIF2004_OptimizeFileSize_Warning_AvoidDuplicativeResultRuleInformation_Text),
             nameof(RuleResources.SARIF2004_OptimizeFileSize_Warning_EliminateLocationOnlyArtifacts_Text),
-            nameof(RuleResources.SARIF2004_OptimizeFileSize_Warning_EliminateIdOnlyRules_Text)
+            nameof(RuleResources.SARIF2004_OptimizeFileSize_Warning_EliminateIdOnlyRules_Text),
+            nameof(RuleResources.SARIF2004_OptimizeFileSize_Warning_PreferRuleId_Text)
         };
 
         public override FailureLevel DefaultLevel => FailureLevel.Warning;
 
+        private string driverGuid;
+
         protected override void Analyze(Run run, string runPointer)
         {
             AnalyzeLocationOnlyArtifacts(run, runPointer);
             AnalyzeIdOnlyRules(run, runPointer);
+
+            this.driverGuid = run.Tool.Driver?.Guid;
+        }
+
+        protected override void Analyze(Result result, string resultPointer)
+        {
+            ReportUnnecessaryAnalysisTarget(result, resultPointer);
+            ReportRuleDuplication(result, resultPointer);
+        }
+
+        private void ReportRuleDuplication(Result result, string resultPointer)
+        {
+            if (result.Rule != null)
+            {
+                if (result.Rule.ToolComponent != null)
+                {
+                    if (result.Rule.ToolComponent.RefersToDriver(this.driverGuid))
+                    {
+                        // {0}: This result uses the 'rule' property to specify the rule
+                        // metadata, but the 'ruleId' property suffices because the rule
+                        // is defined by 'tool.driver'. Prefer 'result.ruleId' because it
+                        // is shorter and just as clear.
+                        LogResult(
+                            resultPointer,
+                            nameof(RuleResources.SARIF2004_OptimizeFileSize_Warning_PreferRuleId_Text));
+                    }
+                    else
+                    {
+                        if (!string.IsNullOrWhiteSpace(result.RuleId) || result.RuleIndex >= 0)
+                        {
+                            // {0}: This result specifies both 'result.ruleId' and 'result.rule'.
+                            // Prefer 'result.ruleId' because it is shorter and just as clear.
```

Comment:
I will provide new strings.

Reply:
New message:

The result at '{0}' uses the 'rule' property to specify the metadata for a rule defined by a tool extension, so it is not necessary also to specify 'ruleId' or 'ruleIndex'. Remove the 'ruleId' and/or 'ruleIndex' properties.

---
In reply to: [455809589](https://github.com/microsoft/sarif-sdk/pull/1995#discussion_r455809589) [](ancestors = 455809589)